annotation/consume_trigger.go

```diff
@@ -71,9 +77,63 @@ type Prestring interface {
 	String() string
 }
 
+// Assignment is a struct that represents an assignment to an expression
+type Assignment struct {
+	LHSExprStr string
+	RHSExprStr string
+	Position   token.Position
+}
+
+func (a *Assignment) String() string {
+	return fmt.Sprintf("`%s` to `%s` at %s", a.RHSExprStr, a.LHSExprStr, a.Position)
+}
+
+// assignmentFlow is a struct that represents a flow of assignments
+type assignmentFlow struct {
+	assignments []Assignment
+}
+
+func (a *assignmentFlow) addEntry(entry Assignment) {
+	a.assignments = append(a.assignments, entry)
+}
+
+func (a *assignmentFlow) String() string {
+	if len(a.assignments) == 0 {
+		return ""
+	}
+
+	// fix point convergence in backprop could mean duplicate entries in this slice, hence we filter out the duplicates here
+	entries := make([]Assignment, 0, len(a.assignments))
+	seen := make(map[string]bool)
+	for _, entry := range a.assignments {
+		if !seen[entry.String()] {
+			seen[entry.String()] = true
+			entries = append(entries, entry)
+		}
+	}
+
```

Comment:
Maybe we can use our own `orderedmap` implementation after PR #66 is merged (I'll try to merge that ASAP) such that we do not have to store a lot of entries and then dedup (i.e., we dedup on the fly).

But taking a step back, would this cause a lot of performance penalty? are there any other ways (maybe not) 🤔

Reply:
Done. Using `orderedmap` now :)

Reply:
Haven't observed any performance penalty as such, but will run a thorough performance experiment to confirm.